post.js

```diff
@@ -54,48 +54,70 @@
-      await initializationP;
-      let copiedBuf = null;
-      if (buf instanceof ArrayBuffer) {
-        copiedBuf = arrayBufferToHeap(buf, 0, buf.byteLength);
-      } else if (buf instanceof TypedArray) {
-        copiedBuf = arrayBufferToHeap(buf.buffer, buf.byteOffset, buf.byteLength);
-      }
-      const channelsPtr = Module._malloc(4);
-      const sampleRatePtr = Module._malloc(4);
-      const outputPtr = Module._malloc(4);
-      const length = decodeMemory(copiedBuf.byteOffset, copiedBuf.byteLength, channelsPtr, sampleRatePtr, outputPtr);
-      if (length < 0) {
-        reject(new Error('stbvorbis decode failed: ' + length));
-        return;
-      }
-      const channels = ptrToInt32(channelsPtr);
-      
-      let data = [];
-      const dataPtrs = ptrToInt32s(ptrToInt32(outputPtr), channels);
-      for (const ptr of dataPtrs) {
-        data.push(ptrToFloat32s(ptr, length));
-      }
-      const result = {
-        data:       data,
-        sampleRate: ptrToInt32(sampleRatePtr),
-      };
+  onmessage = async event => {
+    await initializationP;
+    const buf = event.data.arrayBuffer;
+    let copiedBuf = null;
+    if (buf instanceof ArrayBuffer) {
+      copiedBuf = arrayBufferToHeap(buf, 0, buf.byteLength);
+    } else if (buf instanceof TypedArray) {
+      copiedBuf = arrayBufferToHeap(buf.buffer, buf.byteOffset, buf.byteLength);
+    }
+    const channelsPtr = Module._malloc(4);
+    const sampleRatePtr = Module._malloc(4);
+    const outputPtr = Module._malloc(4);
+    const length = decodeMemory(copiedBuf.byteOffset, copiedBuf.byteLength, channelsPtr, sampleRatePtr, outputPtr);
+    if (length < 0) {
+      postMessage({ id: event.data.id, error: new Error('stbvorbis decode failed: ' + length) });
+      return;
+    }
+    const channels = ptrToInt32(channelsPtr);
+    
+    let data = [];
+    const dataPtrs = ptrToInt32s(ptrToInt32(outputPtr), channels);
+    for (const ptr of dataPtrs) {
+      data.push(ptrToFloat32s(ptr, length));
+    }
+    const result = {
+      id:         event.data.id,
+      data:       data,
+      sampleRate: ptrToInt32(sampleRatePtr),
+    };
 
-      Module._free(copiedBuf.byteOffset);
-      Module._free(channelsPtr);
-      Module._free(sampleRatePtr);
+    Module._free(copiedBuf.byteOffset);
+    Module._free(channelsPtr);
+    Module._free(sampleRatePtr);
 
-      for (const ptr of dataPtrs) {
-        Module._free(ptr);
-      }
-      Module._free(ptrToInt32(outputPtr));
-      Module._free(outputPtr);
+    for (const ptr of dataPtrs) {
+      Module._free(ptr);
+    }
+    Module._free(ptrToInt32(outputPtr));
+    Module._free(outputPtr);
 
-      resolve(result);
-    });
+    postMessage(result, result.data.map(array => array.buffer));
   };
 })(Module);
 
+} // End of the function decodeWorker().
+
+let id = 0;
+const worker = new Worker(URL.createObjectURL(new Blob([ `(${decodeWorker.toString()})();` ], { type: "text/javascript" })));
```

Comment:
Ugh, using `Function.prototype.toString` is fragile (the result is not speced)... How about just defining decoderWorker as a string?

Reply:
sorry I was not understanding "speced"
it's "spaced"?

Reply:
I meant that the result is not defined in the spec

Reply:
https://www.ecma-international.org/ecma-262/6.0/#sec-function.prototype.tostring

it is not enough ? :->

Reply:
The webpage is too big: could you tell me the section number you indicated?

Reply:
Sorry
`19.2.3.5 Function.prototype.toString ( )`

(I linked the just section using `#` on url, but it is not enabled...?)

Reply:
OK i didn't that specification. Let's go with `toString`. Thank you!